File: packages/engine/src/cpus/6502/instructions/rol.test.js
import { test, expect } from 'vitest'
import { BufferMapper } from '../../../buffer_mapper'
import { CPU } from '..'

test('rol acc', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 0x80
  cpu.carry = true
  mmu.data = [0x2a] // ROL
  const g = cpu.run()
  for (let i = 0; i < 2; i++) g.next()
  expect(cpu.acc).toBe(1)
  expect(cpu.carry).toBeTruthy()
})

test('rol zpg', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  mmu.data = [0x26, 0x02, 4] // ROL $02
  const g = cpu.run()
  for (let i = 0; i < 5; i++) g.next()
  expect(mmu.data[2]).toBe(8)
})

test('rol abs', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  mmu.data = [0x2e, 0x02, 0xff] // ROL $ff02
  mmu.data[0xff02] = 4
  const g = cpu.run()
  for (let i = 0; i < 6; i++) g.next()
  expect(mmu.data[0xff02]).toBe(8)
})

File: packages/engine/src/cores/famibob/cpu/instructions/rol.test.zig
const testing = @import("../testing/setup.zig");

test "ROL acc" {
    var cpu = testing.createCPU();
    cpu.acc = 0x80;
    cpu.c = true;
    cpu.write(0, 0x2a);
    for (0..2) |_| {
        cpu.process();
    }
    try testing.expectEqual(true, cpu.c);
    try testing.expectEqual(1, cpu.acc);
}

test "ROL zpg" {
    var cpu = testing.createCPU();
    cpu.acc = 0x80;
    cpu.write(0, 0x26);
    cpu.write(1, 2);
    cpu.write(2, 4);
    for (0..5) |_| {
        cpu.process();
    }
    try testing.expectEqual(8, cpu.read(2));
}

test "ROL abs" {
    var cpu = testing.createCPU();
    cpu.write(0, 0x2e);
    cpu.write(1, 2);
    cpu.write(2, 0xff);
    cpu.write(0xff02, 4);
    for (0..6) |_| {
        cpu.process();
    }
    try testing.expectEqual(8, cpu.read(0xff02));
}
